td1: Adding support for interfacing with TD1 devices

jimmyjon711 opened the pull request:
Feature was requested by AJAX3D(designer of TD1) to add the ability into AFC-Klipper-Add-On to have the ability to automatically collect data from TD1 that are connected inline with filament paths.

So this PR adds a new TD1 module which can automatically connect and interface with multiple TD1s that are connected to a printer. The purpose of this is to be used with filament changers(or standalone) so that color and TD values can be grabbed for each filament that's loaded. This has initially been integrated into a [development branch](https://github.com/jimmyjon711/AFC-Klipper-Add-On/tree/td1) of AFC-Klipper-Add-On and tested by multiple testers. 

Signed-off-by: Jim Madill <jcmadill1@gmail.com>

Arksine commented:
I believe this is a useful addition to Moonraker.  I had some comments about the implementation above.  Thanks.

jimmyjon711 commented:
@Arksine Code has been updated to use `async_serial` and I also moved a lot of the serial logic into `TD1Protocol` class. I probably  made some other mistakes, but please let me know.

I did retest to verify all logic still works like before.

jimmyjon711 commented on moonraker/components/td1.py:
```diff
@@ -0,0 +1,255 @@
+        )
+        self._server.register_endpoint(
+            "/machine/td1_reboot", RequestType.POST, self._handle_reboot
+        )
```
Fixed in latest commit

jimmyjon711 commented:
> I pushed some changes to the `async_serial` wrapper. It can now be instantiated without a config and its possible to use a callback in place of the StreamReader. I notice you have made some changes, if possible I will attempt to do a review later today.
> 
> At a glance one thing I noticed that I missed last time is that most class methods and functions do not include annotations. All methods need to be fully annotated (arguments, except `self`, and return value) for the type checker to work correctly. It isn't necessary to annotate every variable, the type checker will let you know which ones need annotations.

I will go back and do this, something that I glanced over when looking over your contribution docs. I did run `mypy` and it didn't bring up anything that I had to annotate.

Arksine commented:
> I will go back and do this, something that I glanced over when looking over your contribution docs. I did run mypy and it didn't bring up anything that I had to annotate.

FWIW, if a method is not annotated `mypy` will skip type checking on all source code within the method.

Arksine commented on moonraker/components/td1.py:
```diff
@@ -0,0 +1,316 @@
+        self.buffer = ""
+        self.error_state = False
+        self.serial = async_serial.AsyncSerialConnection(config,
+                                                         config.get("baudrate", 115200))
```
It is no longer necessary to pass a config object to AsyncSerial connection.  It will need a server object, a unique name, the port and the baudrate.

Arksine commented on moonraker/components/td1.py:
```diff
@@ -0,0 +1,316 @@
+                self.logger.error(
+                    f"Failed to open port:{port} for TD1 serial {serial_number}")
+
+        except Exception as e:
```
We may not want to handle a bare exception here.  At the very least its necessary to re-raise `asyncio.CancelledError`.

Arksine commented on moonraker/components/td1.py:
```diff
@@ -0,0 +1,316 @@
+                    self.logger.info(f"Hot-plugged TD1 detected: {serial_number}")
+                    try:
+                        await self._start_serial_task(port, serial_number)
+                    except Exception:
```
As with line 278, it is necessary to re-raise `asyncio.CancelledError`.

jimmyjon711 commented on moonraker/components/td1.py:
```diff
@@ -0,0 +1,316 @@
+    async def reboot(self):
+        self.logger.info(f"Rebooting {self.serial_number}")
+        self.enabled = False
+        self.serial_task.cancel()
```
When doing a reset the RP2040 closes the serial port so this is why I am exiting the task. Then when it reconnects `_watch_for_new_devices` will pick it back up and start a new serial task. Then we need to recheck for the error file as a user might not have fixed the original issue.

Arksine replied:
Ok.  I pushed a small change to `async_serial` that will allow you to gracefully exit the serial task so we don't risk the iterator consuming the `ack` from the `change settings` request.  Something like the following should work:

```python
self.serial.set_read_callback(None, force=True)
with contextlib.suppress(asyncio.TimeoutError):
     await asyncio.wait_for(self.serial_task, 1.)
```

The above will feed an EOF to the stream reader, causing the iterator to exit.  A new reader is then instantiated which will be guaranteed to receive the response to the "change settings" command.

jimmyjon711 commented on moonraker/components/td1.py:
```diff
@@ -0,0 +1,316 @@
+
+        self.done_initializing = True
+
+        while self.enabled:
```
This is more me following `wled.py` file as an example and not being as familiar with how async works

jimmyjon711 commented:
@Arksine All comments have been addressed except https://github.com/Arksine/moonraker/pull/1021#discussion_r2366363531. I also removed the commit for `async_serial.py` file. Not sure if you like others to resolve comments but I will keep them unresolved as I would rather you close them to verify that I fixed them correctly. 

Also big thanks for taking your time in helping me through this as I normally don't work with async code like this.

Arksine commented on moonraker/components/td1.py:
```diff
@@ -0,0 +1,337 @@
+
+        except serial.serialutil.SerialException as e:
+            self.logger.error(f"Failed to open serial port {port}: {e}")
+            raise serial.serialutil.SerialException
```
This should be `raise` or `raise e` if you want to re-raise an exception.  It isn't valid to raise the type.

Arksine commented on moonraker/components/td1.py:
```diff
@@ -0,0 +1,337 @@
+            raise serial.serialutil.SerialException
+        except asyncio.CancelledError as e:
+            self.logger.error(f"Task canceled error {e}")
+            raise asyncio.CancelledError
```
Same as line 297.

Arksine commented on moonraker/components/td1.py:
```diff
@@ -0,0 +1,337 @@
+                    self.logger.info(f"Hot-plugged TD1 detected: {serial_number}")
+                    try:
+                        await self._start_serial_task(port, serial_number)
+                    except serial.serialutil.SerialException:
```
There is probably no need for a `try` block here since no exception was handled.  Previously I believe that this was a bare exception and I indicated that `asyncio.CanceledError` needs to be re-raised.  So it would need to be something like:

```python
try:
    await self._start_serial_task(port, serial_number)
except asyncio.CancelledError:
    raise
except Exception:
    # log error here if you would like
    continue
```

Arksine commented on moonraker/components/td1.py:
```diff
@@ -0,0 +1,337 @@
+            await asyncio.wait_for(self.serial.close(), 1)
+        if self.serial_task is not None or not self.serial_task.done():
+            with contextlib.suppress(asyncio.CancelledError):
+                with contextlib.suppress(asyncio.TimeoutError):
```
It isn't necessary to use nested calls to `contextlib.suppress` if you want to suppress multiple exceptions.  You can do the following:

```python
with contextlib.suppress(asyncio.CancelledError, asyncio.TimeoutError):
    await asyncio.wait_for(self.serial_task, 1)
```

Arksine commented on moonraker/components/td1.py:
```diff
@@ -0,0 +1,341 @@
+                self.logger.error(
+                    f"Failed to open port:{port} for TD1 serial {serial_number}")
+
+        except serial.serialutil.SerialException as e:
```
I missed this before, however it appears that `serialutil` is not a valid attribute of `serial`.  This should be `serial.SerialException`.

Arksine commented:
The above comment is the only functional issue that appears to be outstanding.  There are some annotations that need to be cleaned up, however I can take care of that after merging.

After fixing the above I'm prepared to merge once you confirm that the device works as expected.

jimmyjon711 commented:
> The above comment is the only functional issue that appears to be outstanding. There are some annotations that need to be cleaned up, however I can take care of that after merging.
> 
> After fixing the above I'm prepared to merge once you confirm that the device works as expected.

@Arksine Made change and retested, everything looks to be working correctly.

Endpoint results:
<img width="423" height="177" alt="image" src="https://github.com/user-attachments/assets/817fc9ef-bdca-437a-8b4d-21f8138aae99" />

Error when turning on TD1:
<img width="719" height="185" alt="image" src="https://github.com/user-attachments/assets/69d408e1-2777-4990-8f04-86d0b4014ecd" />
